### PR TITLE
Clippy-ify current code

### PR DESCRIPTION
**Pull request by tarquin-the-brave**

### CODE REVIEW

File: build.sh
Comment (rlupton20) [inline build.sh:14]: I wonder if we should add a script/aliases to the floki environment in the build container, with instructions on start, to make it easy for people to check their code isn't going to bump into CI.

For instance I might have an alias `presubmit` which does a build test and clippy, so I can check before pushing. Later we can add e.g. DCO signature checks etc.

Essentially, make it easy to check locally.

**Review by rlupton20 — CHANGES_REQUESTED** [review]
Looks good, one tweak to make it easy for folks to run locally.

File: build.sh
Comment (tarquin-the-brave) [reply build.sh:14]: DCO, you may have a point.  And if the list of things CI does grows there could well be good cause for such a utility.  But right now is there really a cause for something easier than running `cargo`?  There's also `build.sh`.

I'd say not as part of this PR anyway.  I don't want to be the one to put a layer over `cargo`... we've seen that not work so well in the past.

Comment (bossmc) [reply build.sh:14]: You don't really need `--release` on the clippy invocation (though it probably doesn't hurt).

Comment (rlupton20) [reply build.sh:14]: You don't need to worry about the DCO (the CI check is already fast anyway) - I can handle that later. I don't mean anything heavyweight, it's just a shorthand to make it ergonomic to run the 3 or 4 things that you should check before pushing (build, test, clippy, format, later DCO). I think abstract wrappers for general builds are really quite different. Here the environment is fixed, so we can do something reliable. The idea here isn't to replace cargo, but to make it easy for people to know their code is good to go before pushing, without needing to know the ins and outs of what we've decided should fail the CI pipeline. Build and test is fair enough, but format/clippy etc have varied traction, for better or worse, and are much more opinionated. Making it a cargo subcommand isn't out of the question.


I can do this if needed. I was thinking something a little shell function loaded in the floki environment

```
presubmit () {
  cargo build
  cargo check
  cargo fmt --SOME_FLAGS
  cargo clippy
}
```

and emitting a little banner in `init` to let people know its there.

Workflow would look like:

```
----
WELCOME TO FLOKI
Facilities:
  presubmit  -- check code before submitting
----

cargo build
cargo test
...
git commit ...
presubmit
FIX THINGS
git commit --amend ...
presubmit
git push ...
```

`build.sh` would need `dind` which is nuts given we've got everything we need in the environment already. I'd get rid of `build.sh` but travis seems kind of crap and makes dockerized builds gross (maybe there is a better way that I don't know).

Comment (tarquin-the-brave) [reply build.sh:14]: fine.  I've added a `presubmit.sh` & a banner to the `floki.yaml`.  Feel free to merge it and change it if you don't like it/want to change it.  I'm not going to bikeshed over it in this PR.

**Comment by tarquin-the-brave** [comment]
@rlupton20 good to go in?

File: build.sh
Comment (rlupton20) [reply build.sh:14]: Tom, watch the attitude, this comes across poorly. It's not a given right to merge work, and the fact that it's peripheral to the core code, or we're on friendly terms, doesn't entail a drop in standard. Same goes for communications with Max, which I thought were previously misaligned. All suggestions have been well-meaning and constructive, as far as I can tell. It's certainly not "bikeshedding".

I had a look into this, and ignoring the particulars of how you've done it, I think there are more fundamental issues.

- clippy run on it's own doesn't return non-zero on failue. This means it does nothing in the pipeline step at the moment (mistakes happen, but this is basic testing).
- Adding `RUSTFLAGS=-Dwarnings` is probably more aggressive than we want.

I suspect we would want something much more fine grained in terms of CI for this. I think blocking merges on code formatting/clippy is a stupid idea, although well-meaning, because it doesn't reflect the right priorities (it's important to get bug fixes out fast sometimes, not blocked by trivialities). I haven't looked into how to do it, but ideally we have a stage which warns but isn't an absolute failure of the pipeline. Maybe we can use badges and stickers in the README to encourage that stuff to get fixed, and/or a simple checklist on normal MRs so people actually do it.

As such I don't think this MR is quite right without asking for a fair bit more work, and I can't cherry-pick the clippification as a single commit.

My steer is to close this MR, and re-open with just the clippification. @maxdymond, does this seem reasonable to you?

Comment (tarquin-the-brave) [reply build.sh:14]: Sorry. That was a misjudgment on my part. I should have suggested in the direction of not doing the things in CI in this PR. Which as you say, would be more work than what I intended this PR to be I.e. "run clippy".

The initial thing I added to CI was somewhat of an afterthought. I'm happy to concede it's more complicated than I gave it credit for.

**Comment by tarquin-the-brave** [comment]
@rlupton20 I've removed any of the CI additions.  This is only clippy-ifying the current code.  Also re-ran clippy with cargo `1.47.0`.  I don't know why DCO is failing on my last commit.  It was definitely signed. I've tried resigning as per the advice under ["details"](https://github.com/Metaswitch/floki/pull/119/checks?check_run_id=1279740492).  There's a "set DCO to pass", button, but I don't want to just press it if there's something else I need to fix up.

File: build.sh
Comment (rlupton20) [reply build.sh]: That's fine - thinking about CI and how things flow in the project is my problem, so it's fine to not have thought about that. I appreciate the apology, let's keep it professional going forwards. Just to be clear, it's (obviously) fine to disagree with Max or I (we both sometimes take some persuasion) but let's assume good intent.

**Comment by rlupton20** [comment]
Thanks, this looks good.

The DCO failure is for the commit reverting the presubmit script. If you could also squash that would be good (I can do it from here, but since you need to sign anyway).

**Comment by tarquin-the-brave** [comment]
ah ha, that'll be my git config signing commits but not reverts!

**Comment by tarquin-the-brave** [comment]
Anything else before we merge? Do you need to unrequest the change?

**Comment by maxdymond** [comment]
Happy with these changes.

**Pull request merged**
